### PR TITLE
Port fuse_linear from pytorch/tvm

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25974 [quant][graphmode] use whitelist for selecting observed values
* #25625 [quant][graphmode] Fold quantize op into module
* #25624 [quant][graphmode] Add fusion for quantized linear
* **#25623 Port fuse_linear from pytorch/tvm**
* #25505 [quant][graphmode] Support quantizing any methods called

Summary:
Port over fuse_linear pass from pytorch/tvm project, we'll need this
in backend specific quantization pass to match aten::linear and swap
it with quantized linear

Test Plan:
python test/test_jit.py 'TestJit.test_fuse_linear'

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17208890](https://our.internmc.facebook.com/intern/diff/D17208890)

### CODE REVIEW

**Comment by jianyuh** [comment]
Just double check, this is just porting the code from pytorch/tvm, not adding a dependency to pytorch/tvm, right?

**Comment by jerryzh168** [comment]
Yes that is correct

File: torch/csrc/jit/passes/fuse_linear.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *  in later passes
+ */
+#pragma once
+
+#include <torch/csrc/jit/ir.h>
+
+namespace torch {
+namespace jit {
+
+/** \brief Match the at::linear pattern and fuse it into a sigle at::linear
```
Comment (nishantpdce) [inline torch/csrc/jit/passes/fuse_linear.h]: minor typo sigle

**Comment by nishantpdce** [comment]
LGTM

File: torch/csrc/jit/passes/fuse_linear.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+
+    // replace addmm pattern to linear
+    SubgraphRewriter addmm_to_linear;
+    addmm_to_linear.RegisterRewritePattern(addmm_pattern, fused_linear);
+    addmm_to_linear.runOnGraph(graph);
+
+    // replace matmul + add pattern to linear
+    SubgraphRewriter matmuladd_to_linear;
+    matmuladd_to_linear.RegisterRewritePattern(matmul_add_pattern, fused_linear);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/fuse_linear.cpp]: You could register all of the patterns together and then call `runOnGraph` once.

File: torch/csrc/jit/passes/fuse_linear.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return (%res))IR";
+
+    std::string matmul_pattern = R"IR(
+    graph(%input, %weight):
+        %weight_t = aten::t(%weight)
+        %output = aten::matmul(%input, %weight_t)
+        return (%output))IR";
+    std::string fused_linear_bias_none = R"IR(
+    graph(%input, %weight):
+        %bias: Tensor? = prim::Constant()
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/fuse_linear.cpp]: Are you sure this one would run correctly after the transformation?

**Review by ZolotukhinM — APPROVED** [review]
Overall looks good, I have a couple of comments (some here, some inline):

 - When you add a new file, you also need to add it to `tools/build_variables.py`, otherwise internal builds will fail.
 - [Nit] The commit message mentions tvm for no reason - it's not like you ported a pass from TVM, you just implemented a similar pass in PyTorch. It's good to mention the source of origin/inspiration, but I suspect there was not much "porting" involved in this PR.

**Pull request closed without merging**